### PR TITLE
Make docker-push "latest" suffix configurable & support arbitrary docker tag versions

### PR DESCRIPTION
**Pull request by matthewadams**
We don't use `latest` normally, so we needed to make this script more flexible, supporting a configurable `latest` tag version, as well as allowing the caller to specify all additional tag versions that should be pushed.

All behavior is backwards compatible.

### CODE REVIEW

**Comment by endzyme** [comment]
Hi @matthewadams! Thanks for the contribution and I think it's a great one! I'll be reviewing this now.

File: bin/docker-push [hunk 1]
```diff
@@ -9,6 +9,9 @@ if [ -z "$DOCKERTAG" ];                     then echo DOCKERTAG must be set; exi
 if [ -z "$CI_SHA1" ];                       then echo CI_SHA1 must be set; exit 1; fi
 if [ -z "$CI_BUILD_NUM" ];                  then echo CI_BUILD_NUM must be set; exit 1; fi
 
+# support overriding "latest"
+LATEST=${LATEST:-latest}
```
Comment (endzyme) [inline bin/docker-push]: Quick note here - Can we consider a change from `LATEST` to `DOCKER_LATEST_TAG` as the env var? Since these are global I wanted to make sure the vars are clear where their used.

File: bin/docker-push [hunk 2]
```diff
@@ -17,6 +20,11 @@ if [ -z "$CI_REF" ]; then echo CI_BRANCH or CI_TAG must be set; exit 1; fi
 
 CI_REF=$(echo "${CI_REF}" | tr / _)
 
+# support list of additional tag versions delimited by commas by default
+TAG_VERSION_DELIMITER_CHAR="${TAG_VERSION_DELIMITER_CHAR:-,}"
```
Comment (endzyme) [inline bin/docker-push]: Could this also be `DOCKER_TAG_VERSION_DELIMITER_CHAR` to specify what type of tag this is used for in the pipeline?

File: bin/docker-push [hunk 3]
```diff
@@ -41,8 +49,9 @@ if [ "$DOCKER_BUILD_CACHE_FROM" == "available" ]; then
   done <<< "$(grep -i '^FROM.* AS ' ${DOCKERFILE} | awk '{print $4}')"
 fi
 
-echo "Pushing ${DOCKERTAG}:latest as ${CI_SHA1}, ${CI_REF}, ${CI_BUILD_NUM} to ${EXTERNAL_REGISTRY_BASE_DOMAIN}/${REPOSITORY_NAME}"
-
-tag_push "${DOCKERTAG}:latest" "${EXTERNAL_REGISTRY_BASE_DOMAIN}/${REPOSITORY_NAME}:${CI_SHA1}"
-tag_push "${DOCKERTAG}:latest" "${EXTERNAL_REGISTRY_BASE_DOMAIN}/${REPOSITORY_NAME}:${CI_REF}"
-tag_push "${DOCKERTAG}:latest" "${EXTERNAL_REGISTRY_BASE_DOMAIN}/${REPOSITORY_NAME}:build_${CI_BUILD_NUM}"
+for TAG_VERSION in $(echo "$TAG_VERSIONS" | tr "$TAG_VERSION_DELIMITER_CHAR" ' '); do
```
Comment (endzyme) [inline bin/docker-push]: I like the extensibility here! Although I generally avoid the use of arrays in bash, I think they make sense here due to how we use `*.config` files for pipeline configurations. I'll give an example below:

The `development.config` below shows an example configuration with your current implementation.
```bash
# development.config - the pipeline parameters for rok8s-scripts development pipeline jobs
NAMESPACE='development'
SOPS_SECRETS=('development/some-secret-file')
HELM_CHARTS=('charts/myapp')
HELM_RELEASE_NAMES=('myapp-dev')
HELM_VALUES=('development/myapp')

# EXAMPLE IMPLEMENTATION FOR OVERRIDE
LATEST=something-not-latest
SOME_ADDTNL_TAG=another-tag # this could have been reduced into the line below but I wanted to be consistent with env var reference
TAG_VERSIONS="$CI_SHA1$TAG_VERSION_DELIMITER_CHAR$CI_REF$TAG_VERSION_DELIMITER_CHAR$BUILD_TAG_PREFIX$CI_BUILD_NUM$TAG_VERSION_DELIMITER_CHAR$SOME_ADDTNL_TAG"
```

Since we're already using bash arrays in configurations for other downstream scripts, we could change the workflow to have a default set of tags that are built and be able to provide additional tags if wanted, OR override the default tags. See the config example below.

```bash
# development.config - the pipeline parameters for rok8s-scripts development pipeline jobs
NAMESPACE='development'
SOPS_SECRETS=('development/some-secret-file')
HELM_CHARTS=('charts/myapp')
HELM_RELEASE_NAMES=('myapp-dev')
HELM_VALUES=('development/myapp')

# DIFFERENT EXAMPLE IMPLEMENTATION FOR OVERRIDE
DOCKER_LATEST_TAG=something-not-latest
ADDITIONAL_DOCKER_PUSH_TAGS=('another-tag' 'yet-another')
```

The above example give the user the ability to add tags without having to restate default tags they may still want. The script implementation can still have `TAG_VERSIONS` but instead of a strict override we could say `TAG_VERSIONS+=(${ADDITIONAL_DOCKER_PUSH_TAGS[@]})` to append the array of additional tags (if defined). 

I'm definitely open to a conversation here and my intent is not to block but provide a different user experience for people defining parameters in the `*.config` files. If you want to chat on a video call we can also talk through this too!

File: bin/docker-push (hunk 1 above)
Comment (matthewadams) [reply bin/docker-push]: sure that's easy

File: bin/docker-push (hunk 2 above)
Comment (matthewadams) [reply bin/docker-push]: sure

File: bin/docker-push (hunk 3 above)
Comment (matthewadams) [reply bin/docker-push]: That's a pretty easy fix too, although I'd personally rather folks either take the default tags or provide their own.  Seems overkill to me to allow them to specify additional tags.  Once they want to take control, then they should just take control.  A first impl I'll push will be to use a bash array, then you can consider whether we should allow _additional_ tags to be specified versus just allowing _tags_ to be specified.

Comment (matthewadams) [reply bin/docker-push]: BTW:
> LATEST=something-not-latest
SOME_ADDTNL_TAG=another-tag # this could have been reduced into the line below but I wanted to be consistent with env var reference
TAG_VERSIONS="$CI_SHA1$TAG_VERSION_DELIMITER_CHAR$CI_REF$TAG_VERSION_DELIMITER_CHAR$BUILD_TAG_PREFIX$CI_BUILD_NUM$TAG_VERSION_DELIMITER_CHAR$SOME_ADDTNL_TAG"

If you notice, `TAG_VERSION_DELIMITER_CHAR` defaults to `,`, so I think the above would become
```
LATEST=something-not-latest
SOME_ADDTNL_TAG=another-tag # this could have been reduced into the line below but I wanted to be consistent with env var reference
TAG_VERSIONS="$CI_SHA1,$CI_REF,$BUILD_TAG_PREFIX$CI_BUILD_NUM,$SOME_ADDTNL_TAG"
```
I didn't really want to make the tag delimiter be configurable lest it be overkill, but it _is_ the more general solution, so I went with that; I didn't want someone to come along later & say, "Hey, _I_ like to delimit with colons"...

Comment (matthewadams) [reply bin/docker-push]: Oh, hell.  I went ahead & just added support for specifying _additional_ tags, too.  Also changed `TAG_VERSIONS` to `DOCKER_TAG_VERSIONS` to be consistent.  The script now expects `DOCKER_TAG_VERSIONS` and `ADDITIONAL_DOCKER_TAG_VERSIONS` to be arrays.

**Comment by matthewadams** [comment]
ok now?

File: bin/docker-push [hunk 4]
```diff
@@ -41,8 +57,9 @@ if [ "$DOCKER_BUILD_CACHE_FROM" == "available" ]; then
...
 
-echo "Pushing ${DOCKERTAG}:latest as ${CI_SHA1}, ${CI_REF}, ${CI_BUILD_NUM} to ${EXTERNAL_REGISTRY_BASE_DOMAIN}/${REPOSITORY_NAME}"
-
-tag_push "${DOCKERTAG}:latest" "${EXTERNAL_REGISTRY_BASE_DOMAIN}/${REPOSITORY_NAME}:${CI_SHA1}"
-tag_push "${DOCKERTAG}:latest" "${EXTERNAL_REGISTRY_BASE_DOMAIN}/${REPOSITORY_NAME}:${CI_REF}"
-tag_push "${DOCKERTAG}:latest" "${EXTERNAL_REGISTRY_BASE_DOMAIN}/${REPOSITORY_NAME}:build_${CI_BUILD_NUM}"
+for DOCKER_TAG_VERSION in ${ALL_DOCKER_TAG_VERSIONS[@]}; do
+  if [ "${DOCKER_LATEST_TAG}" != "${DOCKER_TAG_VERSION}" ]; then
+    echo "Pushing ${DOCKERTAG}:${DOCKER_LATEST_TAG} as ${EXTERNAL_REGISTRY_BASE_DOMAIN}/${REPOSITORY_NAME}:${DOCKER_TAG_VERSION}"
+    tag_push "${DOCKERTAG}:${DOCKER_LATEST_TAG}" "${EXTERNAL_REGISTRY_BASE_DOMAIN}/${REPOSITORY_NAME}:${DOCKER_TAG_VERSION}"
```
Comment (endzyme) [inline bin/docker-push]: This looks great! I am looking at where else `:latest` might be inferred and I think this might have `./bin/docker-build` implications. After crawling through this a bit I see that the `$DOCKERTAG` variable is not such a great name since it looks to be the image _and_ (potentially) tag name!

If someone has defined the tag name in the `$DOCKERTAG` variable then we'll hit conditions where this line could read `image-name:sometag:$DOCKER_LATEST_TAG` (which would break).

The root of the problem here is that `$DOCKERTAG` isn't named very well at all. To accommodate this how would you feel about adding in a check to assure we don't have a `:` in the `$DOCKERTAG` variable.

```bash
# Adding override convention of DOCKERTAG
DOCKER_LATEST_TAG=${DOCKER_LATEST_TAG:-latest}
if [[ "${DOCKERTAG}" =~ ":" ]]; then
  echo "The \$DOCKERTAG (${DOCKERTAG}) var contains a tag name. Skipping adding \$DOCKER_LATEST_TAG (${DOCKER_LATEST_TAG})."
else
  echo "The \$DOCKERTAG (${DOCKERTAG}) var does not contain a default tag. Appending \$DOCKER_LATEST_TAG ($DOCKER_LATEST_TAG)."
  DOCKERTAG="${DOCKERTAG}:${DOCKER_LATEST_TAG}"
fi
```
(Feel free to trash the above block, I just want to make sure we're not breaking implementations where people are using a tag in `$DOCKERTAG` field.)

I could be mistaken too - I'm jumping into this part of `rok8s-scripts` and it's a little hard to tell where some of these variables are used :-\ -- thanks again for your patience and contribution.

Comment (endzyme) [reply bin/docker-push]: I am also asking internally to see if anyone who originally wrote this piece has some insight into intent of `DOCKERTAG`

Comment (endzyme) [reply bin/docker-push]: Ok - after tracing through the history of this repo, it would appear that `docker-push` has always used the `docker tag ${DOCKERTAG}:latest` convention. This means no one should have ever been able to set `DOCKERTAG=image:tag` and have `./bin/docker-push` work properly. This leads me to believe all implementations that currently exist will _not_ have a tag name in `DOCKERTAG` variable.

Since the `docker-build` script needs `DOCKERTAG` to do its builds, we should at least add the same block to set the `DOCKER_LATEST_TAG` in `./bin/docker-build`. I also see `:latest` used in `./bin/docker-save`, so this should also be replaced with the `DOCKER_LATEST_TAG` convention.

So, long story short: We should be fine to approve & merge as long as `./bin/docker-build` and `./bin/docker-save` are updated to utilize the new `DOCKER_LATEST_TAG` override tag.

Sound ok?

Comment (matthewadams) [reply bin/docker-push]: Done.  See https://github.com/reactiveops/rok8s-scripts/pull/213/commits/8db98c995a77875a6ef5c3907f20316cc62ee407

**Comment by endzyme** [comment]
@matthewadams This looks awesome! Thanks for the perseverance and follow-through! I'm going to merge this into v7.21.0-branch which will be released fairly shortly.

**Comment by matthewadams** [comment]
👍🏼
